contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql

```diff
@@ -2577,6 +2577,254 @@ AS 'babelfishpg_tsql', 'update_user_catalog_for_guest';
+  	        WHEN ss.prokind = 'p' THEN 56
+  	        ELSE sdit.ss_data_type
+  	      END
+  	    WHEN st.is_table_type = 1 THEN 0
+  	    ELSE sdit.ss_data_type
+  	  END
+  	AS sys.tinyint) AS SS_DATA_TYPE
+  , CAST(ss.proname AS sys.sysname) AS original_procedure_name
+  FROM 
+  ( 
+    -- CTE to query procedures related to bbf
+     WITH bbf_proc AS (
+   	   SELECT
+         p.proname as proname,
+         p.proargnames as proargnames,
+         p.proargmodes as proargmodes,
+         p.prokind as prokind,
+         p.proretset as proretset,
+         p.prorettype as prorettype,
+         p.pronamespace as pronamespace,
+         p.proallargtypes as proallargtypes,
+         p.proargtypes as proargtypes
+       FROM 
+         pg_proc p
+       WHERE (
+	     p.pronamespace in (select schema_id from sys.schemas union all select oid from pg_namespace where nspname = 'sys')
+	     AND (pg_has_role(p.proowner, 'USAGE') OR has_function_privilege(p.oid, 'EXECUTE'))
+	     AND (p.probin like '{%typmod_array%}'-- typmod array exists for those that are user-defined
+	       OR p.proname like 'sp\_%' -- filter out internal babelfish-specific procs
+	       OR p.proname like 'xp\_%'
+	       OR p.proname like 'dm\_%'
+	       OR p.proname like 'fn\_%'))
+	)
+	       
+     SELECT -- Selects all parameters (input and output), but NOT return values
+       p.proname as proname,
+       p.proargnames as proargnames,
+       p.proargmodes as proargmodes,
+       p.prokind as prokind,
+       p.proretset as proretset,
+       p.prorettype as prorettype,
+       p.pronamespace as pronamespace,
+       information_schema._pg_expandarray(
+       COALESCE(p.proallargtypes,
+         CASE 
+           WHEN p.prokind = 'f' THEN (CAST(p.proargtypes AS oid[]))
+           ELSE CAST(p.proargtypes AS oid[])
+         END
+       )) AS x
+    FROM bbf_proc p
+      
+    UNION ALL
+      
+    SELECT -- Selects all return values
+      p.proname as proname,
+      p.proargnames as proargnames,
+      p.proargmodes as proargmodes,
+      p.prokind as prokind,
+      p.proretset as proretset,
+      p.prorettype as prorettype,
+      p.pronamespace as pronamespace,
+      NULL AS x -- null value indicates that we are retrieving the return values of the proc/func
+    FROM bbf_proc p
+  ) ss
+  INNER JOIN pg_catalog.pg_namespace ns ON ns.oid = ss.pronamespace
+  LEFT JOIN sys.babelfish_namespace_ext bne ON ns.nspname = bne.nspname 
+  LEFT JOIN sys.databases d ON d.database_id = bne.dbid
+  LEFT JOIN sys.types st ON ((ss.x IS NOT NULL AND st.user_type_id = (ss.x).x) OR (ss.x IS NULL AND st.user_type_id = ss.prorettype))
+  -- Because spt_datatype_info_table does contain user-defind types and their names,
+  -- the join below allows us to retrieve the system type name on what the user-defined type was based on.
+  LEFT JOIN sys.spt_datatype_info_table sdit ON sdit.type_name = (SELECT name FROM sys.types WHERE user_type_id = st.system_type_id)
```

Comment:
Sorry that I am still confused with this join and the use of subquery in join condition. Could you please give me an example of this?

Reply:
So I need sys.spt_datatype_info_table because it contains the ODBC information about the system datatypes. However, this table is hardcoded so any user-defined types would have this information. Furthermore, this table only contains the data-type names (no OID). 

So if I create a type like this: ` CREATE TYPE my_udt FROM INT`   I get a newly generated user_type_id (unique to this type) and a system_type_id (which corresponds to INT) in `sys.types`. 

If I try the suggested join condition `sdit.type_name = st.name`. This wouldn't work because `my_udt` does not exist in `sys.spt_datatype_info_table`, only the names of the system types do. 

The only way to get the ODBC information about the user-defined type is to find the name of the datatype in `sys.types` that has `my_udt`'s `system_type_id` as their `user_type_id`.